rust/src/dns/detect.rs

```diff
@@ -189,3 +263,43 @@ mod test {
         ));
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn parse_rcode_good() {
+        assert_eq!(
+            parse_rcode("1"),
+            Ok(DetectDnsRcode {
+                rcode: 1
+            })
+        );
+        assert_eq!(
+            parse_rcode("123"),
+            Ok(DetectDnsRcode {
+                rcode: 123
+            })
+        );
+        assert_eq!(parse_rcode(""), Err(()));
+        assert_eq!(parse_opcode("asdf"), Err(()));
+    }
+
+    #[test]
+    fn test_match_rcode() {
+        assert!(match_rcode(
+            &DetectDnsRcode {
+                rcode: 0,
+            },
+            0b0000_0000_0000_0000,
+        ));
+
+        assert!(match_rcode(
+            &DetectDnsRcode {
+                rcode: 4,
+            },
+            0b0010_0000_0000_0000,
```

Comment:
`rcode` still be the last 4 bits.  So an `rcode` of 4 will be `0b0000_0000_0000_0100`.